ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/BlsToExecutionOperationPool.java

```diff
@@ -0,0 +1,198 @@
+  }
+
+  @Override
+  public void subscribeOperationAdded(
+      final OperationAddedSubscriber<SignedBlsToExecutionChange> subscriber) {
+    this.subscribers.subscribe(subscriber);
+  }
+
+  @Override
+  public SszList<SignedBlsToExecutionChange> getItemsForBlock(final BeaconState stateAtBlockSlot) {
+    return getItemsForBlock(stateAtBlockSlot, operation -> true, operation -> {});
+  }
+
+  @Override
+  public SszList<SignedBlsToExecutionChange> getItemsForBlock(
+      final BeaconState stateAtBlockSlot,
+      final Predicate<SignedBlsToExecutionChange> filter,
+      final Consumer<SignedBlsToExecutionChange> includedItemConsumer) {
+    final SszListSchema<SignedBlsToExecutionChange, ?> schema =
+        slotToSszListSchemaSupplier.apply(stateAtBlockSlot.getSlot());
+
+    // Note that iterating through all items does not affect their access time so we are effectively
+    // evicting the oldest entries when the size is exceeded as we only ever access via iteration.
+    final Collection<SignedBlsToExecutionChange> sortedViableOperations = operations.values();
+    final List<SignedBlsToExecutionChange> selected = new ArrayList<>();
+    for (final SignedBlsToExecutionChange item : sortedViableOperations) {
+      if (!filter.test(item)) {
+        continue;
+      }
+      final int validatorIndex = getValidatorIndex(item);
+      if (operationValidator.validateForBlockInclusion(stateAtBlockSlot, item).isEmpty()) {
+        selected.add(item);
+        includedItemConsumer.accept(item);
+        if (selected.size() == schema.getMaxLength()) {
+          break;
+        }
+      } else {
+        // The item is no longer valid to be included in a block so remove it from the pool.
+        operations.remove(validatorIndex);
+      }
+    }
+    return schema.createFromElements(selected);
+  }
+
+  @Override
+  public SafeFuture<InternalValidationResult> addLocal(final SignedBlsToExecutionChange item) {
+    final int validatorIndex = getValidatorIndex(item);
+    if (operations.containsKey(validatorIndex)) {
+      return SafeFuture.completedFuture(rejectForDuplicatedMessage(validatorIndex))
+          .thenPeek(result -> validationReasonCounter.labels(result.code().toString()).inc());
+    }
+    return add(item, false);
+  }
+
+  @Override
+  public SafeFuture<InternalValidationResult> addRemote(final SignedBlsToExecutionChange item) {
+    final int validatorIndex = getValidatorIndex(item);
+    if (operations.containsKey(validatorIndex)) {
+      return SafeFuture.completedFuture(rejectForDuplicatedMessage(validatorIndex))
+          .thenPeek(result -> validationReasonCounter.labels(result.code().toString()).inc());
+    }
+    return add(item, true);
+  }
+
+  @Override
+  public void addAll(final SszCollection<SignedBlsToExecutionChange> items) {
+    items.forEach(
+        item -> {
+          final int validatorIndex = getValidatorIndex(item);
+          operations.putIfAbsent(validatorIndex, item);
```

Comment:
QUESTION

Do you think we should consider re-broadcasting the messages in this case? I don't think we were doing it before, but maybe it makes sense?

Reply:
nope, totally doesnt make sense here imo. can discuss why if you like...